### DISCUSSION TITLE
Note on `pdfjs-dist` security vulnerability

### DISCUSSION THREAD

**Discussion by wojtekmaj** [wojtekmaj/react-pdf#1786] [Announcements]
> [!IMPORTANT]  
> - Update to 9.0.0 or later if you can
> - Update to 7.7.3, 8.0.2 if you can't - then you're safe, even though automated security audits may still fail (longer explanation below).

## What happened?

GHSA-wgrm-67xf-hhpq security advisory has been published in `pdfjs-dist`, a key dependency of ours, noting the use of `eval`, potentially allowing malicious PDF files to execute unrestricted attacker-controlled JavaScript in the context of the hosting domain.

Important: this vulnerability is **only** exploitable when `isEvalSupported` option is set to `true` (default).

## How we addressed the vulnerability in React-PDF?

- `isEvalSupported` option is now forcefully being set to `false`, completely removing the attack vector. **This patch has been released in versions 7.7.3 and 8.0.2.**
- PDF.js has been updated to a fully non-vulnerable version in version 9.0.0.

## Security audit still fail! What can I do?

* 😴   GHSA-wgrm-67xf-hhpq (CVE-2024-4367) can be safely dismissed/ignored. This security advisory means you're (indirectly) using `pdfjs-dist` that's _potentially_ exploitable. This is expected if you're using React-PDF version older than 9.0.0.
* ⚠️ GHSA-87hq-q4gp-9wr4 (CVE-2024-34342) should **not** be ignored. This security advisory means you're using React-PDF version that is still missing the patch making `pdfjs-dist` vulnerability not exploitable.

## Why didn't you just update `pdfjs-dist` to the latest version immediately?

It was the long term plan, but it was too early to leave v7/v8 users behind. Updating `pdfjs-dist` to the latest version was a breaking change. This meant leaving v7 and v8 users behind. A quick, non-breaking patch release was, I believe, welcome by most.

Furthermore, `pdfjs-dist` versions earlier than 4.3.136 included changes that would break many setups, even the most modern ones.

**Comment by Scc33** [comment]
Thanks for the great detail!

**Comment by Ryanv030** [comment]
Appreciate the detail laid out here. However this puts us in an awkward spot. I don't want to turn off my pre-commit hook that checks for security vulnerabilities and if I keep the pre-commit hook on this will keep failing until `pdfjs-dist` is updated to version 4.2.67. 

Is there a timetable for the longterm plan or would you recommend looking for new solutions outside of react-pdf?

**Reply by wojtekmaj** [reply]
Our timeline is flexible yet closely linked to the release of this PR: https://github.com/mozilla/pdf.js/pull/18051, which will significantly improve the upgradability of the next breaking version of React-PDF which will include the definite fix.

**Reply by matthew-mogford-veriforce** [reply]
Seeing a lot of positive comments and Green on that pdf.js PR!
Fingers crossed it gets in soon.

**Reply by stevelizcano** [reply]
Just got merged!

**Reply by Hcrab2336** [reply]
For anyone else running into the npm audit issue, our devops team decided to run the npm audit step in a powershell script and grep for the exact output that pdfjs-dist gives when it fails alone and exit code 0 or 1 accordingly. It's not ideal but it's a temporary fix that keeps us running while still checking for vulnerabilities. I don't really know the specifics but hopefully this helps others until the fix is fully implemented.

**Reply by Ryanv030** [reply]
Appreciate the update @wojtekmaj!

**Reply by wojtekmaj** [reply]
@Ryanv030 @mattmogford-alcumus @stevelizcano @Hcrab2336 Good news! v9.0.0 is released and vulnerability is now fully fixed.

**Reply by Ryanv030** [reply]
Thank you @wojtekmaj!

**Comment by shye0000** [comment]
Thx for this announcement, it really clears things up for my team! We use this package and the date picker too, they kill perfectly our pain points of UI developement, well maintained with great communications! Hope for the best of this security issue!

**Reply by wojtekmaj** [reply]
@shye0000 Good news! v9.0.0 is released and vulnerability is now fully fixed.

**Reply by shye0000** [reply]
@wojtekmaj Super good news! Thank you so much!

**Comment by cpernsteineribm** [comment]
Finally https://github.com/mozilla/pdf.js/pull/18051 got released https://github.com/mozilla/pdf.js/releases/tag/v4.3.136 🚀

**Reply by wojtekmaj** [reply]
@cpernsteineribm v9.0.0 is released and includes pdfjs-dist 4.3.136. :)

**Comment by TobiTenno** [comment]
migrating to v9 for the security fix is a pretty heavy lift if you're not already on esm builds. any chance to port _just_ the security fix back to v8 so we can get the fix and update to ESM in our own time?

**Reply by wojtekmaj** [reply]
There's absolutely no way. We have: 
v7 & 8: vulnerable, CJS-based PDF.js 3 + vulnerability mitigation (100% safe)
v9: non-vulnerable, ESM-based PDF.js 4, no mitigation necessary

**Reply by TobiTenno** [reply]
I see. Apologies for the extra noise.

**Comment by lucianoaraus** [comment]
Thanks for the detail man! Something very important [here](https://github.com/wojtekmaj/react-pdf/issues/321#issuecomment-2205920661), this was mandatory for me. I couldn't move to v9 until I changed this:

pdfjs.GlobalWorkerOptions.workerSrc = `https://cdnjs.cloudflare.com/ajax/libs/pdf.js/${pdfjs.version}/pdf.worker.min.js`

**To this**:
pdfjs.GlobalWorkerOptions.workerSrc = `https://cdnjs.cloudflare.com/ajax/libs/pdf.js/${pdfjs.version}/pdf.worker.min.mjs`

This was the error that I had before the update btw:

> Warning: Error: Setting up fake worker failed: "Failed to fetch dynamically imported module: https://cdnjs.cloudflare.com/ajax/libs/pdf.js/4.8.69/pdf.worker.min.js". Error Component Stack

**Reply by TobiTenno** [reply]
for anyone who was as confused as me for a solid 20s, the change is the extension (`js` -> `mjs`).

**Reply by wojtekmaj** [reply]
It's best to read the entire upgrade guide here: https://github.com/wojtekmaj/react-pdf/wiki/Upgrade-guide-from-version-8.x-to-9.x

**Comment by maiiiiiiii** [comment]
what about folks using v6 because of older browser compatibility? :/

**Reply by wojtekmaj** [reply]
v6 won't receive any further updates, but you can disable the problematic option yourself.

**Comment by uttejcorpay** [comment]
@wojtekmaj Recently i upgraded my company's project's `react-pdf` to v9 to resolve this security advisory. Though we have resolved the issue with latest version, there is a strange behavior i observed. When i upload a PDF that contains a basic (`alert()`) script embedded via OpenActions and when `isEvalSupported:false` passed to options on `Document`, i could still able to see the alert being executed after the upload. 

Did i miss anything here?
